lightning/src/onion_message/messenger.rs

```diff
@@ -1146,7 +1146,57 @@ where
 			log_trace!(self.logger, "Dropping event {:?}: buffer full", event);
 			return
 		}
-		pending_events.push(event);
+		pending_events.push_back(event);
+	}
+
+	/// Processes any events asynchronously using the given handler.
+	///
+	/// Note that the event handler is called in the order each event was generated, however
+	/// futures are polled in parallel for some events to allow for parallelism where events do not
+	/// have an ordering requirement.
+	///
+	/// See the trait-level documentation of [`EventsProvider`] for requirements.
+	pub async fn process_pending_events_async<Future: core::future::Future<Output = ()> + core::marker::Unpin, H: Fn(Event) -> Future>(
+		&self, handler: H
+	) {
+		let mut events = VecDeque::new();
+		core::mem::swap(&mut *self.pending_events.lock().unwrap(), &mut events);
+		let mut futures = Vec::with_capacity(events.len());
+		for (node_id, recipient) in self.message_recipients.lock().unwrap().iter_mut() {
+			if let OnionMessageRecipient::PendingConnection(_, addresses, _) = recipient {
+				if let Some(addresses) = addresses.take() {
+					futures.push(Some(handler(Event::ConnectionNeeded { node_id: *node_id, addresses })));
+				}
+			}
+		}
+		// We process events in parallel, but we want to complete `OnionMessageIntercepted` events
+		// prior to `OnionMessagePeerConnected` ones.
+		for _ in 0..events.len() {
+			let next_event = events.pop_front();
+			match next_event {
+				Some(Event::OnionMessageIntercepted { .. }) => {
+					futures.push(Some(handler(next_event.unwrap())));
+				},
```

Comment:
Could this parallelism affect how users structure their database storage for intercepted OMs? Just wondering if some extra docs could be helpful.